### ISSUE TITLE
font-charset

### ISSUE THREAD

**Issue by jack9267** [mikke89/RmlUi#36]
The font interface libRocket as a whole relies on "font-charset". If this applies to RmlUi is there a possibility of removing font-charset workaround in order to simply generate glyphs as needed or at least let me use my own better font renderer.

I mostly say this as I want to use emoji and russian and many other languages (depending on user choice).

https://mikke89.github.io/RmlUiDoc/pages/localisation.html

**Comment by mikke89** [comment]
Thanks for the suggestion, generating glyphs as needed would be a good enhancement and we should aim for emoji support. 

See also the suggestion for a font provider interface in #10. @viciious: I think we should keep this in mind when designing a new font interface, any thoughts?

**Comment by jack9267** [comment]
It would be great to use my own font system (it has draw functions) saves memory and other bits.
Although using halfbuzz would be better for font layout.
FontProvider would likely be the way to go.

**Comment by jack9267** [comment]
Ideally you could generate new sets of 512x512 or whatever textures to contain glyphs as needed. New texture once old one is filled up.

**Comment by viciious** [comment]
> Thanks for the suggestion, generating glyphs as needed would be a good enhancement and we should aim for emoji support.
> 
> See also the suggestion for a font provider interface in #10. @viciious: I think we should keep this in mind when designing a new font interface, any thoughts?

What we do in our engine is lazy-load the rendered characters in our own 512x512 or 1024x1024 atlases in the GenerateString call. The entire ASCII range is pre-loaded the first time the font is touched. This avoids both introducing new API calls and the font-charset hack.

**Comment by mikke89** [comment]
I think this is the way to go for the internal font provider as well, so we can remove the font-charset.

Emoji support will be more work though. At least we need to add rgb font textures and figure out how to best deal with possibly multiple textures for the given text. Also, I think we should make all internal text encoding UTF8, currently we use UCS-2 internally and to my understanding this cannot represent emojis among other things.

**Comment by jack9267** [comment]
Represent text as UTF8 and store char32_t (4 byte text) for emoji support.

Free type handles emojis fine alongs it’s updated with very little effort! I support emoji even.

**Comment by mikke89** [comment]
That sounds good! How do you deal with rgb emojis, do you store these in a separate texture from the main glyphs?

**Comment by jack9267** [comment]
It works the same as any other character the emoji the diff is there’s like a colour channel iirk I can look later and explain it better

**Comment by jack9267** [comment]
PNG emoji (such as google noto) works a bit different though and it’s not much effort still! More or less blitting

**Comment by mikke89** [comment]
Started working on this in the 'font' branch. Very much a work-in-progress currently, but we have loading of (pre-specified) emojis up and running. Moved all internal characters to UTF-8 encoding.

Will implement lazy loading of unicode characters, and a "fallback"-font. Colored and bitmap emojis are lower priority for now at least.

**Comment by viciious** [comment]
From my experience, colored emojis are a chore because they need to be really high res to look good and then there's the problem of finding a suitable emoji pic resolution for the given font size, so that they look good next to each other.
Alternatively, they should not be rasterized at all and use the vector data for rt rendering on the gpu but that's probably not going to be worth the effort, unless your game is all about chatting...

**Comment by mikke89** [comment]
Alright, we'll see when we get to it. It seems FreeType makes it quite easy to rasterize colored (vector) glyphs, so it might actually not be too much effort. I can imagine bitmap emojis are more tricky with the points you raise.

The main issue right now seems to be dealing with the font effects while updating the glyphs. Makes everything quite a bit more complicated.

**Comment by viciious** [comment]
Yeah.. I just ditched font effects altogether in my font engine implementation.

**Comment by mikke89** [comment]
I've merged the font branch into develop now. The main changes/features:

- `font-charset` property is gone: The font interface now loads new characters as needed, which also work when font-effects are applied. Fallback fonts can be set so that unknown characters are loaded from them.
- Grayscale emojis are supported, no longer limited to 16 bit code points, have a look at the 'demo' sample for some examples. Colored emojis may or may not be supported in the future.
- The API and internals are now using UTF-8 strings directly, the old UCS-2 strings are ditched completely, but we have a UTF-16 converter for some assistance when working with the Windows API.
- The font interface API is slightly modified, some functions renamed to make it slightly more helpful to understand.
- The internals of the default font engine has had a major overhaul, simplifying a lot of code, and removing the BitmapFont provider.
- Instead, a custom font engine interface has been implemented for bitmap fonts in the bitmapfont sample. A quick example of how to create your own font interface.
- The CMake flag `NO_FONT_INTERFACE_DEFAULT` now removes the default font engine files from the source file list. Macro guards are no longer needed and have been removed.

Let me know if you have any feedback on the changes :)
